File: src/agent.ts
/** Main orchestrator for the TAM/SOM assessment pipeline. */

import { classifyIndustry } from "./analysis/industry.js";
import { estimateTopDown } from "./analysis/top-down.js";
import { estimateBottomUp } from "./analysis/bottom-up.js";
import { reconcileTam } from "./analysis/reconcile.js";
import { analyzeCompetitiveLandscape } from "./analysis/competitive.js";
import { estimateSom } from "./analysis/som.js";
import { getSettings, type Settings } from "./config.js";
import { LLMClient } from "./llm/client.js";
import { methodologySummaryPrompt, SYSTEM_PROMPT } from "./llm/prompts.js";
import type { CompanyInput, FinalReport } from "./models.js";
import { WebSearcher } from "./research/web-search.js";
import { formatUsd, formatPct } from "./util/format.js";

export type ProgressCallback = (
  step: number,
  totalSteps: number,
  label: string,
  detail?: string,
) => void;

export class TAMAgent {
  private llm: LLMClient;
  private searcher: WebSearcher;
  public onProgress?: ProgressCallback;

  constructor(settings?: Settings) {
    const s = settings ?? getSettings();
    this.llm = new LLMClient(s);
    this.searcher = new WebSearcher(s);
  }

  private emit(step: number, label: string, detail?: string) {
    if (this.onProgress) {
      this.onProgress(step, 6, label, detail);
    }
    const prefix = `\x1b[1;34mStep ${step}/6:\x1b[0m`;
    if (detail) {
      console.log(`${prefix} ${label}\n  ${detail}`);
    } else {
      console.log(`${prefix} ${label}`);
    }
  }

  async run(company: CompanyInput): Promise<FinalReport> {
    // Step 1: Industry Classification
    this.emit(1, "Classifying industry...");
    const industry = await classifyIndustry(this.llm, company);
    this.emit(
      1,
      "Industry classified",
      `NAICS ${industry.naics_code}: ${industry.naics_description} (confidence: ${formatPct(industry.confidence)})`,
    );

    // Step 2: Top-down TAM
    this.emit(2, "Estimating TAM (top-down)...");
    const { estimate: tamTd } = await estimateTopDown(
      this.llm,
      this.searcher,
      company,
      industry,
    );
    this.emit(
      2,
      "Top-down TAM estimated",
      `${formatUsd(tamTd.value_usd)} (confidence: ${formatPct(tamTd.confidence)})`,
    );

    // Step 3: Bottom-up TAM
    this.emit(3, "Estimating TAM (bottom-up)...");
    const { estimate: tamBu } = await estimateBottomUp(
      this.llm,
      this.searcher,
      company,
      industry,
    );
    this.emit(
      3,
      "Bottom-up TAM estimated",
      `${formatUsd(tamBu.value_usd)} (confidence: ${formatPct(tamBu.confidence)})`,
    );

    // Step 4: Reconcile
    this.emit(4, "Reconciling TAM estimates...");
    const tamConsensus = await reconcileTam(
      this.llm,
      company.name,
      company.geography,
      tamTd,
      tamBu,
    );
    this.emit(
      4,
      "Consensus TAM calculated",
      `${formatUsd(tamConsensus.value_usd)} (confidence: ${formatPct(tamConsensus.confidence)})`,
    );

    // Step 5: Competitive Landscape
    this.emit(5, "Analyzing competitive landscape...");
    const { landscape: competitive } = await analyzeCompetitiveLandscape(
      this.llm,
      this.searcher,
      company,
      industry,
    );
    this.emit(
      5,
      "Competitive landscape analyzed",
      `${competitive.market_concentration} concentration, ~${competitive.total_competitors_estimated} competitors`,
    );

    // Step 6: SOM
    this.emit(6, "Estimating SOM...");
    const som = await estimateSom(this.llm, company, tamConsensus, competitive);
    this.emit(
      6,
      "SOM estimated",
      `${formatUsd(som.value_usd)} (${som.market_share_pct.toFixed(1)}% share, confidence: ${formatPct(som.confidence)})`,
    );

    // Methodology summary
    console.log("\x1b[2mGenerating methodology summary...\x1b[0m");
    const methodology = await this.llm.textQuery(
      methodologySummaryPrompt({
        name: company.name,
        industry: company.industry,
        naics_code: industry.naics_code,
        geography: company.geography,
        tam_td: formatUsd(tamTd.value_usd),
        td_conf: formatPct(tamTd.confidence),
        tam_bu: formatUsd(tamBu.value_usd),
        bu_conf: formatPct(tamBu.confidence),
        tam_consensus: formatUsd(tamConsensus.value_usd),
        consensus_conf: formatPct(tamConsensus.confidence),
        som: formatUsd(som.value_usd),
        som_share: som.market_share_pct.toFixed(1),
        som_conf: formatPct(som.confidence),
        num_competitors: competitive.top_competitors.length,
        concentration: competitive.market_concentration,
      }),
      SYSTEM_PROMPT,
    );

    console.log(
      `\x1b[2mToken usage: ${this.llm.totalInputTokens.toLocaleString()} input, ` +
        `${this.llm.totalOutputTokens.toLocaleString()} output\x1b[0m`,
    );

    return {
      company,
      industry,
      tam_top_down: tamTd,
      tam_bottom_up: tamBu,
      tam_consensus: tamConsensus,
      competitive_landscape: competitive,
      som,
      methodology_notes: methodology,
      generated_at: new Date().toISOString(),
    };
  }
}

File: tam_agent/agent.py
"""Main orchestrator for the TAM/SOM assessment pipeline."""

from __future__ import annotations

from rich.console import Console

from tam_agent.analysis.bottom_up import estimate_bottom_up
from tam_agent.analysis.competitive import analyze_competitive_landscape
from tam_agent.analysis.industry import classify_industry
from tam_agent.analysis.reconcile import reconcile_tam
from tam_agent.analysis.som import estimate_som
from tam_agent.analysis.top_down import estimate_top_down
from tam_agent.config import Settings
from tam_agent.llm.client import LLMClient
from tam_agent.llm.prompts import METHODOLOGY_SUMMARY, SYSTEM_PROMPT
from tam_agent.models import CompanyInput, FinalReport
from tam_agent.research.web_search import WebSearcher

console = Console()


class TAMAgent:
    """Orchestrates the multi-step TAM/SOM assessment pipeline."""

    def __init__(self, settings: Settings | None = None) -> None:
        self.settings = settings or Settings()
        self.llm = LLMClient(self.settings)
        self.searcher = WebSearcher(self.settings)

    def run(self, company: CompanyInput) -> FinalReport:
        """Execute the full TAM/SOM assessment pipeline."""

        # Step 1: Industry Classification
        console.print("\n[bold blue]Step 1/6:[/] Classifying industry...", highlight=False)
        industry = classify_industry(self.llm, company)
        console.print(
            f"  NAICS {industry.naics_code}: {industry.naics_description} "
            f"(confidence: {industry.confidence:.0%})"
        )

        # Step 2: Top-down TAM
        console.print("\n[bold blue]Step 2/6:[/] Estimating TAM (top-down)...", highlight=False)
        tam_td, _, _ = estimate_top_down(self.llm, self.searcher, company, industry)
        console.print(
            f"  Top-down TAM: ${tam_td.value_usd:,.0f} "
            f"(confidence: {tam_td.confidence:.0%})"
        )

        # Step 3: Bottom-up TAM
        console.print("\n[bold blue]Step 3/6:[/] Estimating TAM (bottom-up)...", highlight=False)
        tam_bu, _, _ = estimate_bottom_up(self.llm, self.searcher, company, industry)
        console.print(
            f"  Bottom-up TAM: ${tam_bu.value_usd:,.0f} "
            f"(confidence: {tam_bu.confidence:.0%})"
        )

        # Step 4: Reconcile
        console.print("\n[bold blue]Step 4/6:[/] Reconciling TAM estimates...", highlight=False)
        tam_consensus = reconcile_tam(
            self.llm, company.name, company.geography, tam_td, tam_bu
        )
        console.print(
            f"  Consensus TAM: ${tam_consensus.value_usd:,.0f} "
            f"(confidence: {tam_consensus.confidence:.0%})"
        )

        # Step 5: Competitive Landscape
        console.print("\n[bold blue]Step 5/6:[/] Analyzing competitive landscape...", highlight=False)
        competitive, _, _ = analyze_competitive_landscape(
            self.llm, self.searcher, company, industry
        )
        console.print(
            f"  Market concentration: {competitive.market_concentration} "
            f"({competitive.total_competitors_estimated} competitors)"
        )

        # Step 6: SOM
        console.print("\n[bold blue]Step 6/6:[/] Estimating SOM...", highlight=False)
        som = estimate_som(self.llm, company, tam_consensus, competitive)
        console.print(
            f"  SOM: ${som.value_usd:,.0f} ({som.market_share_pct:.1f}% share, "
            f"confidence: {som.confidence:.0%})"
        )

        # Generate methodology notes
        console.print("\n[dim]Generating methodology summary...[/]")
        methodology = self.llm.text_query(
            prompt=METHODOLOGY_SUMMARY.format(
                name=company.name,
                industry=company.industry,
                naics_code=industry.naics_code,
                geography=company.geography,
                tam_td=tam_td.value_usd,
                td_conf=f"{tam_td.confidence:.0%}",
                tam_bu=tam_bu.value_usd,
                bu_conf=f"{tam_bu.confidence:.0%}",
                tam_consensus=tam_consensus.value_usd,
                consensus_conf=f"{tam_consensus.confidence:.0%}",
                som=som.value_usd,
                som_share=som.market_share_pct,
                som_conf=f"{som.confidence:.0%}",
                num_competitors=len(competitive.top_competitors),
                concentration=competitive.market_concentration,
            ),
            system=SYSTEM_PROMPT,
        )

        console.print(
            f"\n[dim]Token usage: {self.llm.total_input_tokens:,} input, "
            f"{self.llm.total_output_tokens:,} output[/]"
        )

        return FinalReport(
            company=company,
            industry=industry,
            tam_top_down=tam_td,
            tam_bottom_up=tam_bu,
            tam_consensus=tam_consensus,
            competitive_landscape=competitive,
            som=som,
            methodology_notes=methodology,
        )
